File: routes/userRoutes.js
const express = require("express");
const router = express.Router();
const userController = require("../controllers/userController");
const multer = require("multer");
const upload = multer({ storage: multer.memoryStorage() });

/**
 * @swagger
 * /users:
 *   get:
 *     summary: Get a list of users
 *     description: Retrieve a list of users from the database.
 *     responses:
 *       200:
 *         description: Successful response with a list of users.
 */

/**
 * @swagger
 * /register:
 *   post:
 *     summary: Register a new user
 *     description: Create a new user account by providing necessary details.
 *     responses:
 *       201:
 *         description: User successfully registered.
 */
router.post("/register", userController.createUser);

/**
 * @swagger
 * /confirm-email:
 *   post:
 *     summary: Confirm user email
 *     description: Confirms the user's email address.
 *     responses:
 *       200:
 *         description: Email successfully confirmed.
 */
router.post("/confirm-email", userController.confirmEmail);

/**
 * @swagger
 * /login:
 *   post:
 *     summary: User login
 *     description: Logs in a user with the provided credentials.
 *     responses:
 *       200:
 *         description: User successfully logged in.
 */
router.post("/login", userController.loginUser);

/**
 * @swagger
 * /logout:
 *   post:
 *     summary: User logout
 *     description: Logs out the currently authenticated user.
 *     responses:
 *       200:
 *         description: User successfully logged out.
 */
router.post("/logout", userController.logout);

/**
 * @swagger
 * /profile:
 *   get:
 *     summary: Get user profile
 *     description: Retrieve the profile details of the authenticated user.
 *     responses:
 *       200:
 *         description: Successfully retrieved user profile.
 */
router.get("/profile", userController.getProfile);

/**
 * @swagger
 * /profile:
 *   patch:
 *     summary: Update user profile
 *     description: Update the profile details of the authenticated user.
 *     responses:
 *       200:
 *         description: User profile updated successfully.
 */
router.patch("/profile", userController.updateProfile);

/**
 * @swagger
 * /profile:
 *   delete:
 *     summary: Delete user profile
 *     description: Delete the profile of the authenticated user.
 *     responses:
 *       200:
 *         description: User profile deleted successfully.
 */
router.delete("/profile", userController.deleteProfile);

/**
 * @swagger
 * /reset-password:
 *   post:
 *     summary: Reset password
 *     description: Sends a reset password email to the user.
 *     responses:
 *       200:
 *         description: Reset password email sent successfully.
 */
router.post("/reset-password", userController.initiateRecovery);

/**
 * @swagger
 * /change-password/{id}:
 *   patch:
 *     summary: Change password
 *     description: Change the password for the user with the given ID.
 *     parameters:
 *       - in: path
 *         name: id
 *         required: true
 *         description: ID of the user.
 *         schema:
 *           type: string
 *     responses:
 *       200:
 *         description: Password changed successfully.
 */
router.patch("/change-password/:param", userController.changePassword);

/**
 * @swagger
 * /verify-reset-link:
 *   get:
 *     summary: Verify reset link
 *     description: Verifies the reset password link sent to the user.
 *     responses:
 *       200:
 *         description: Reset link verified successfully.
 */
router.get("/verify-reset-link", userController.verifyResetLink);




/**
 * @swagger
 * /mfa-settings:
 *   post:
 *     summary: Save MFA settings
 *     description: Save the multi-factor authentication settings for a user.
 *     responses:
 *       200:
 *         description: MFA settings saved successfully.
 */
router.post("/mfa-settings", userController.saveMfaSettings);

/**
 * @swagger
 * /verify-mfa:
 *   post:
 *     summary: Verify MFA code
 *     description: Verify the multi-factor authentication code provided by the user.
 *     responses:
 *       200:
 *         description: MFA code verified successfully.
 */
router.post("/verify-mfa", userController.verifyMfaCode);

/**
 * @swagger
 * /resend-code/{email}:
 *   get:
 *     summary: Resend confirmation code
 *     description: Resend the confirmation code to the user's email.
 *     parameters:
 *       - in: path
 *         name: email
 *         required: true
 *         description: User's email address.
 *         schema:
 *           type: string
 *     responses:
 *       200:
 *         description: Confirmation code resent successfully.
 */
router.get("/resend-code/:email", userController.resendConfirmationCode);

// Define the file upload route
router.post("/upload", upload.single("file"), userController.uploadFile);
router.post("/setup-2fa", userController.setUp2FA);
// Route to request a magic link
router.post('/request-magic-link', userController.sendMagicLink);
router.post("/verify-2fa", userController.verify2FA);

// Route to verify magic link when the user clicks the link
router.get('/magic-link', userController.verifyMagicLink);
router.post('/recover-account', userController.initiateRecovery);
router.post('/recovery-verify', userController.verifyRecovery);
router.post('/generate-private-key', userController.addPrivateAndPublicKey);
router.get('/completeWebAuthRegisteration', userController.createWebAuthRegisteration);
router.post('/complete-webauth-register', userController.completeWebAuthRegisteration);
router.post('/webauthn/complete-authenticate', userController.completeWebAuthnAuthentication);
router.post('/verify-security-pin', userController.verifysecurityPin)
module.exports = router;

File: routes/user_routes.py
from fastapi import APIRouter, UploadFile, File
from controllers.user_controller import UserController

router = APIRouter(prefix="/users", tags=["Users"])

@router.post("/register", response_model=dict, status_code=201)
async def create_user(data: dict):
    """Register a new user."""
    return await UserController.create_user(data)

@router.post("/confirm-email", response_model=dict)
async def confirm_email(data: dict):
    """Confirm user email."""
    return await UserController.confirm_email(data)

@router.post("/login", response_model=dict)
async def login_user(data: dict):
    """User login."""
    return await UserController.login_user(data)

@router.post("/logout", response_model=dict)
async def logout():
    """User logout."""
    return await UserController.logout()

@router.get("/profile", response_model=dict)
async def get_profile():
    """Retrieve user profile."""
    return await UserController.get_profile()

@router.patch("/profile", response_model=dict)
async def update_profile(data: dict):
    """Update user profile."""
    return await UserController.update_profile(data)

@router.delete("/profile", response_model=dict)
async def delete_profile():
    """Delete user profile."""
    return await UserController.delete_profile()

@router.post("/reset-password", response_model=dict)
async def initiate_recovery(data: dict):
    """Send reset password email."""
    return await UserController.initiate_recovery(data)

@router.patch("/change-password/{param}", response_model=dict)
async def change_password(param: str, data: dict):
    """Change user password."""
    return await UserController.change_password(param, data)

@router.get("/verify-reset-link", response_model=dict)
async def verify_reset_link(token: str, email: str):
    """Verify reset password link."""
    return await UserController.verify_reset_link(token, email)

@router.post("/mfa-settings", response_model=dict)
async def save_mfa_settings(data: dict):
    """Save MFA settings."""
    return await UserController.save_mfa_settings(data)

@router.post("/verify-mfa", response_model=dict)
async def verify_mfa_code(data: dict):
    """Verify MFA code."""
    return await UserController.verify_mfa_code(data)

@router.get("/resend-code/{email}", response_model=dict)
async def resend_confirmation_code(email: str):
    """Resend confirmation code."""
    return await UserController.resend_confirmation_code(email)

@router.post("/upload", response_model=dict)
async def upload_file(file: UploadFile = File(...)):
    """Upload user file."""
    return await UserController.upload_file(file)

@router.post("/setup-2fa", response_model=dict)
async def setup_2fa(data: dict):
    return await UserController.setup_2fa(data)

@router.post("/request-magic-link", response_model=dict)
async def send_magic_link(data: dict):
    return await UserController.send_magic_link(data)

@router.post("/verify-2fa", response_model=dict)
async def verify_2fa(data: dict):
    return await UserController.verify_2fa(data)

@router.get("/magic-link", response_model=dict)
async def verify_magic_link(token: str, email: str):
    return await UserController.verify_magic_link(token, email)

@router.post("/recover-account", response_model=dict)
async def initiate_recovery(data: dict):
    return await UserController.initiate_recovery(data)

@router.post("/recovery-verify", response_model=dict)
async def verify_recovery(data: dict):
    return await UserController.verify_recovery(data)

@router.post("/generate-private-key", response_model=dict)
async def add_private_and_public_key(data: dict):
    return await UserController.add_private_and_public_key(data)

@router.get("/completeWebAuthRegisteration", response_model=dict)
async def create_web_auth_registration():
    return await UserController.create_web_auth_registration()

@router.post("/complete-webauth-register", response_model=dict)
async def complete_web_auth_registration(data: dict):
    return await UserController.complete_web_auth_registration(data)

@router.post("/webauthn/complete-authenticate", response_model=dict)
async def complete_web_authn_authentication(data: dict):
    return await UserController.complete_web_authn_authentication(data)

@router.post("/verify-security-pin", response_model=dict)
async def verify_security_pin(data: dict):
    return await UserController.verify_security_pin(data)